### PR TITLE
Fix to ensure the checkpoint states are saved in a common filepath with deepspeed

### LINKED ISSUE #12864: optimizer states are saved in different folders
## 🐛 Bug

For a step, the model states, optim states are split (saved) into different folders.
```
$ tree .
.
├── T5-monitor-epoch=0-step=200-val_loss=0.1618-val_ppl=1.1909-f1_score=0.3613.ckpt
│   ├── checkpoint
│   │   ├── mp_rank_00_model_states.pt
│   │   ├── zero_pp_rank_0_mp_rank_00_optim_states.pt
│   │   ├── zero_pp_rank_1_mp_rank_00_optim_states.pt
│   │   ├── zero_pp_rank_3_mp_rank_00_optim_states.pt
│   │   ├── zero_pp_rank_4_mp_rank_00_optim_states.pt
│   │   └── zero_pp_rank_6_mp_rank_00_optim_states.pt
│   ├── latest
│   └── zero_to_fp32.py
├── T5-monitor-epoch=0-step=200-val_loss=0.1618-val_ppl=1.1909-f1_score=0.6021.ckpt
│   └── checkpoint
│       ├── zero_pp_rank_2_mp_rank_00_optim_states.pt
│       ├── zero_pp_rank_5_mp_rank_00_optim_states.pt
│       └── zero_pp_rank_7_mp_rank_00_optim_states.pt
├── T5-monitor-epoch=1-step=568-val_loss=0.0352-val_ppl=1.0372-f1_score=0.8835.ckpt
│   ├── checkpoint
│   │   ├── mp_rank_00_model_states.pt
│   │   ├── zero_pp_rank_0_mp_rank_00_optim_states.pt
│   │   ├── zero_pp_rank_1_mp_rank_00_optim_states.pt
│   │   ├── zero_pp_rank_3_mp_rank_00_optim_states.pt
│   │   ├── zero_pp_rank_4_mp_rank_00_optim_states.pt
│   │   └── zero_pp_rank_6_mp_rank_00_optim_states.pt
│   ├── latest
│   └── zero_to_fp32.py
└── T5-monitor-epoch=1-step=568-val_loss=0.0352-val_ppl=1.0372-f1_score=0.9235.ckpt
    └── checkpoint
        ├── zero_pp_rank_2_mp_rank_00_optim_states.pt
        ├── zero_pp_rank_5_mp_rank_00_optim_states.pt
        └── zero_pp_rank_7_mp_rank_00_optim_states.pt

8 directories, 22 files
```

I use `DeepSpeed` strategy, which runs in 8 A100 cards.

Related code:
```py
def validation_step(self, batch, batch_idx):
    inputs = self.create_inputs(batch)
    outputs = self(inputs)
    loss = outputs["loss"]
    ppl = torch.exp(loss)
    metrics = {"val/loss": loss, "val/ppl": ppl.detach()}
    self.log_dict(metrics, prog_bar=True, logger=True, on_epoch=True, batch_size=inputs["batch_size"][0])
    if self.metric is not None:
        pred_seqs = self.generate(inputs)
        metrics["val_pred_seqs"] = pred_seqs
        metrics["gold_seqs"] = inputs["gold_seqs"]
    return metrics

def validation_epoch_end(self, validation_step_outputs):
    if self.metric is not None:
        pred_seqs = []
        gold_seqs = []
        for batch_out in validation_step_outputs:
            pred_seqs.extend(batch_out["val_pred_seqs"])
            gold_seqs.extend(batch_out["gold_seqs"])
        eval_results = self.metric.evaluate(pred_seqs, gold_seqs)
        self.log_dict(eval_results, prog_bar=True, logger=True)
```

### Environment

* CUDA:
        - GPU:
                - A100-SXM4-40GB
                - A100-SXM4-40GB
                - A100-SXM4-40GB
                - A100-SXM4-40GB
                - A100-SXM4-40GB
                - A100-SXM4-40GB
                - A100-SXM4-40GB
                - A100-SXM4-40GB
        - available:         True
        - version:           11.3
* Packages:
        - numpy:             1.20.1
        - pyTorch_debug:     False
        - pyTorch_version:   1.10.2+cu113
        - pytorch-lightning: 1.6.0
        - tqdm:              4.63.1
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - 
        - processor:         x86_64
        - python:            3.7.10
        - version:           #1 SMP Fri Mar 19 10:07:22 CST 2021

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #12864

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @awaelchli @ananthsub @ninginthecloud @rohitgr7 @SeanNaren @akihironitta

### CODE REVIEW

File: tests/strategies/test_deepspeed_strategy.py [hunk 1]
```diff
@@ -1269,13 +1269,19 @@ def test_deepspeed_with_meta_device(tmpdir):
 def test_deepspeed_multi_save_same_filepath(tmpdir):
```
Comment (rohitgr7) [inline tests/strategies/test_deepspeed_strategy.py:1269]: update this test instead of adding another one since deepspeed tests are expensive on ci.

File: tests/strategies/test_deepspeed_strategy.py [hunk 2]
```diff
@@ -1284,6 +1290,10 @@ def test_deepspeed_multi_save_same_filepath(tmpdir):
         enable_model_summary=False,
     )
     trainer.fit(model)
+
+    expected = ["last.ckpt", "epoch=1_step=2_grank=0.0.ckpt"]
+    assert set(expected) == set(os.listdir(trainer.checkpoint_callback.dirpath))
+
     ckpt_path = os.path.join(trainer.checkpoint_callback.dirpath, "last.ckpt")
     expected = ["latest", "zero_to_fp32.py", "checkpoint"]
     assert set(expected) == set(os.listdir(ckpt_path))
```
Comment (carmocca) [inline tests/strategies/test_deepspeed_strategy.py:1299]: (for consistency)

```suggestion
    expected = {"latest", "zero_to_fp32.py", "checkpoint"}
    assert expected == set(os.listdir(ckpt_path))
```

File: tests/strategies/test_deepspeed_strategy.py [hunk 3]
```diff
@@ -1269,13 +1269,19 @@ def test_deepspeed_with_meta_device(tmpdir):
...
+            return super().training_step(*args, **kwargs)
+
+    model = CustomModel()
     trainer = Trainer(
         default_root_dir=tmpdir,
         strategy="deepspeed",
         accelerator="gpu",
         devices=2,
-        callbacks=[ModelCheckpoint(save_top_k=1, save_last=True)],
+        callbacks=[ModelCheckpoint(filename="{epoch}_{step}_{grank}", save_top_k=1, save_last=True)],
```
Comment (carmocca) [inline tests/strategies/test_deepspeed_strategy.py:1284]: Is epoch and step relevant to this test?

```suggestion
        callbacks=[ModelCheckpoint(filename="{grank}", save_top_k=1, save_last=True)],
```

Comment (carmocca) [reply tests/strategies/test_deepspeed_strategy.py:1284]: And `save_last`?

```suggestion
        callbacks=[ModelCheckpoint(filename="{epoch}_{step}_{grank}", save_top_k=1)],
```

**Review by carmocca — APPROVED** [review]

File: tests/strategies/test_deepspeed_strategy.py (hunk 3 above)
Comment (rohitgr7) [reply tests/strategies/test_deepspeed_strategy.py:1284]: `save_last` is not required now.

I'd keep `step` to ensure the new checkpoint path is not the same as the old one and the old one is deleted too.

**Review by otaj — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]
